N3 & rdf-js: Follow the spec more closely by adding union types

LaurensRietveld opened the pull request:
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/rdfjs/representation-task-force
- [ ] Increase the version number in the header if appropriate.

This PR make the typings of rdf-js more accurate by using union types for `Term` and for the positions in quad statements.
I've also applied the same method for N3, considering the strong similarities between the typings of both projects

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -13,87 +13,104 @@ import * as RDF from "rdf-js";
     constructor(iri: string);
+    id: string;
+    toJSON(): {};
+    equals(other: Term): boolean;
```
`RDF.Term` must be accepted here. (also for all following `equals`)

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -13,87 +13,104 @@ import * as RDF from "rdf-js";
+    static subclass(type: any): void;
     language: string;
-    datatype: RDF.NamedNode;
+    datatype: NamedNode;
```
`RDF.NamedNode`

LaurensRietveld replied:
This datatype is part of the return value of the `literal` function in the factory. The N3 factory returns a n3 NamedNode. Any reason why you'd like this on to be of type `RDF.NamedNode`?

rubensworks replied:
The reason is because `literal` in the datafactory should also take an `RDF. NamedNode` as datatype argument. This is because any kind of named node should be allowed to define the datatype, not only those that were created by N3.

LaurensRietveld replied:
afaik the literal function would _always_ return a an N3 NamedNode (see https://github.com/rdfjs/N3.js/blob/master/lib/N3DataFactory.js#L279 , where if the datatype is a `RDF.NamedNode` it would be re-instantiated as a new N3.NamedNode).
Looking at this, I think the datatype input argument of the `literal` datafactory function should be changed from N3.NamedNode to RDF.NamedNode

LaurensRietveld replied:
I'm assuming this is resolved? To clarify: this is the typing of the N3 Literal, which is created by the N3 datafactory. As a result, it'll always have a N3 NamedNode as datatype

LaurensRietveld replied:
A, missed your :+1:  here. Considering it resolved ;)

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -13,87 +13,104 @@ import * as RDF from "rdf-js";
+    static subclass(type: any): void;
 }
 
+export type Quad_Subject = NamedNode | BlankNode | Variable;
```
Same comment here as in the RDFJS typings. The N3 implementation doesn't restrict the term type for S, P, O and G. So the typings should also allow any term type to be used for triple/quad components.

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -152,17 +169,17 @@ export interface WriterConstructor {
 export const Writer: WriterConstructor;
 
 export interface N3Writer {
```
The changes to the input args are not needed I think, only those to the return types.

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -13,87 +13,104 @@ import * as RDF from "rdf-js";
 export type QuadPredicate = (result: Quad) => boolean;
 
-export type OTerm = RDF.Term | string | null;
+export type OTerm = Term | string | null;
```
This should also stay and `RDF.Term`.

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -212,12 +229,12 @@ export interface StoreOptions {
 }
 
 export namespace Util {
```
Same comment here regarding I/O types.

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -177,33 +194,33 @@ export interface N3StreamWriter extends NodeJS.ReadWriteStream, RDF.Source {}
     // match, removeMatches and deleteGraph are missing for full RDF.Store adherence
     remove(stream: stream.Stream): EventEmitter;
 }
 export interface StoreConstructor {
```
Same comment here regarding I/O types.

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -177,33 +194,33 @@ export interface N3StreamWriter extends NodeJS.ReadWriteStream, RDF.Source {}
 
 export interface N3Store extends RDF.Sink {
```
Same comment here regarding I/O types.

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -13,87 +13,104 @@ import * as RDF from "rdf-js";
 
 export type Logger = (message?: any, ...optionalParams: any[]) => void;
 
 export interface BlankTriple {
```
Changes not needed here either.

LaurensRietveld commented:
> The Quad_Subject, ... types are a good idea, but may be too restrictive for some use cases. (For example for generalized RDF-mode in JSON-LD, see comments inline) So I think it's better to not use these types for anything, but it makes sense to at least define them (and explain them in jsdoc).

Good point about the generalized RDF mode. Still, that would leave the option open of either:
- Adding blank node as possible term for sub, pred, object and graphs
- Using generics so users are able to choose the specificity.

Reason I'm saying this is that the more the typescript spec follows the standard (whether that's the rdf-js spec or the w3c), the better.

> The changes to the N3 typings are too restrictive. I would undo all changes to the input arguments of all functions, because now only N3 terms can be used as inputs to these functions, while any RDFJS term should be accepted. The output arguments of all functions can remain (but perhaps not those to the DataFactory, as I'm not sure that TS will still see it as a proper implementation of the RDF.DataFactory (to be checked))

Good point, I've updated the input arguments accordingly.
About the datafactory: added a test for that, and typescript doesnt seem to complain.

I've pushed the new version based using your feedback. Let's discuss the remaining ones

rubensworks commented:
> Using generics so users are able to choose the specificity.

W3C-compliance by default, and non-compliance through generics sounds like a good idea. It's just important that these generics don't become too cumbersome to use. Did you have anything specific in mind for designing these generic types?

It might make sense to mention this issue at the RDFJS gitter/github/mailinglist, so that other people can share their opinion on this.

LaurensRietveld commented:
> W3C-compliance by default, and non-compliance through generics sounds like a good idea. It's just important that these generics don't become too cumbersome to use. Did you have anything specific in mind for designing these generic types?

It would mean having default generics on the parser, writer, store and DataFactory.quad. When e.g. instantiating a parser, writer or store you can change the term types per position. This is only on instantiation (i.e. you don't have to specify typings for the `store.addQuad` func).
These default generics won't be a usability issue for the standard-compliant code.
Those who don't follow the standards would have to specify the generics each time they instantiate a parser, writer, store or datafactory.  (which sounds like a good distribution of labour to me)

> It might make sense to mention this issue at the RDFJS gitter/github/mailinglist, so that other people can share their opinion on this.

I'll post a link on gitter so other can pitch in when they want

LaurensRietveld commented on types/n3/index.d.ts:
```diff
@@ -13,87 +13,104 @@ import * as RDF from "rdf-js";
-    function quad(subject: RDF.Term, predicate: RDF.Term, object: RDF.Term, graph?: RDF.Term): RDF.Quad;
+    function namedNode(value: string): NamedNode;
+    function blankNode(value?: string): BlankNode;
+    function literal(value: string | number, languageOrDatatype?: string | NamedNode): Literal;
```
```suggestion
    function literal(value: string | number, languageOrDatatype?: string | RDF.NamedNode): Literal;
```
I think this should be a RDF.NamedNode

LaurensRietveld commented:
@rubensworks I think we resolved everything except the s/p/o/g definitions. I'll just update the PR with the generics, so we can see the impact on the definitions

fletcher91 commented:
I've been working on rdflib.js types for some time now, they'll differ from spec, but it might be useful for additional reference on implementation details. https://github.com/fletcher91/link-lib/blob/master/%40types/rdflib/index.d.ts

The `sI` properties can be ignored since they are my own addition, I [normalize each triple entering the store](https://github.com/fletcher91/link-lib/blob/master/src/RDFStore.ts#L48) for a significant performance boost.

LaurensRietveld commented:
Thanks @fletcher91 for the link. 
@rubensworks : I added the generics in the latest commit. The typescript def is more complex now, but from a usage perspective nothing should have changed. 
I still feel a bit unease though implementing these generics to allow people _not_ to follow the standard, but guess that ship has sailed when this non-normative definition made it's way into the text of the rdf 1.1 standard.

LaurensRietveld commented:
About the failing test: when we agree on this interface, I'll update the rdf-data-model (min typescript version to 2.3) as well

rubensworks commented:
@LaurensRietveld I'll be available again on Tuesday, I'll try to review it then (or the day after).

fletcher91 commented:
> including Literals, in the subject position
> Makes the system simpler too, and simpler to explain.

The data model might be simpler, but it sorta defeats the purpose of a type system. It can get muddy when trying to implement code as well, you'd have to type guard every subject before calling methods or ending the function;

```typescript
const s = new Subject(NS.ex('a'), NS.ex('a'), Literal.fromValue('c'));
$rdf.fetcher.load(s.subject) // Error TS2345: arg of type `NN|BN|Literal` not assignable to NamedNode
```
One must either cast the type, defeating type-safety;
```typescript
$rdf.fetcher.load(s.subject as NamedNode)
```
or introduce type guards, which introduce additional but essentially unnecessary code paths;
```typescript
if (s.subject.termType !== 'NamedNode') {
  throw new Error('Only a NamedNode can be loaded')
}
$rdf.fetcher.load(s.subject as NamedNode)
```

If one were to allow this, I'd advise to use either generics or [discriminated unions](https://basarat.gitbooks.io/typescript/docs/types/discriminated-unions.html) so people that need the functionality can toggle it;

```typescript
type SubjectTypes = NamedNode | BlankNode | Variable
type GeneralizedTypes = SubjectTypes & Literal | Collection
type 

class Statement<ST = SubjectTypes, PT = NamedNode, OT= Term> {
  public subject: ST;
  public predicate: PT;
  public object: OT;
}

class Formula<ST = SubjectTypes, PT = NamedNode, OT = Term> {
  add(s: ST, p: PT, o: OT): Statement<ST, PT, OT>
}

// Where one now can override the type to be generalized (or any other constraint set)
const generalizedFormula = new Formula<GeneralizedTypes, GeneralizedTypes, GeneralizedTypes>();
```

Or with discriminated union (might be a bit more confusion for end-usage);
```typescript
type SubjectTypes = NamedNode | BlankNode | Variable
type GeneralizedTypes = ClosedSubjectTypes & Literal | Collection;

class Statement {
  public readonly stmtType: string;
  public object: NamedNode | BlankNode | Collection // | etc
}
class ClosedStatement {
  public readonly stmtType: "closed"
  public subject: SubjectTypes
  public predicate: NamedNode
}
class GeneralizedStatement {
  public readonly stmtType: "generalized"
  public subject: GeneralizedTypes
  public predicate: GeneralizedTypes // Allowing all terms according to RDF TR
}

// And just let the user and library authors decide which to use
class Formula<ST = > {
  public readonly graphType: string
}
class ClosedFormula {
  public readonly graphType: "closed"
  add(s: SubjectTypes // etc
}
class GeneralizedFormula {
   public readonly graphType: "generalized"
   public add(s: GeneralizedSubjectTypes // etc
}
```

rubensworks commented:
I did some more thinking on this, and based on @timbl's comment, I think it might be better to make _liberal_ RDF the default, and _strict_ RDF optional.

One main argument in favor of this is that the RDFJS DataFactory interface allows any S/P/O term type. It does not say —for example— that when a literal should be provided as subject, an error should be thrown.

It appears that the RDFJS interfaces are designed to be a generalized form of RDF, where strict RDF is optional, but requires additional checks by the developer if needed. So I think it makes sense to adhere to this as much as possible.

LaurensRietveld commented:
> One main argument in favor of this is that the RDFJS DataFactory interface allows any S/P/O term type. It does not say —for example— that when a literal should be provided as subject, an error should be thrown.

@rubensworks The spec mentions e.g. `subject: the subject, which is a NamedNode, BlankNode or Variable.` I get that this isn't a closed set though (see https://github.com/rdfjs/representation-task-force/issues/121). But shouldn't the _default_ generics correspond as much as possible to the documentation?

We're now supporting arbitrary terms in different statement positions using generics (see [this test](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29836/files#diff-df2307d9e1728f97eb5e7803fbab2fabR80)). I.e., the approach described by @timbl is supported

I think the current state of the PR reflects the rdfjs spec and documentation (http://rdf.js.org/#quad-interface) very well: 
- A statement position must be a term
- The _default_ allowed terms for the statement positions are taken from the rdfjs spec as well
- Deviating from the statement positions definitions for e.g. generalized rdf is possible in the PR presented here (just as it is in the rdfjs spec)

rubensworks commented on types/n3/index.d.ts:
```diff
@@ -16,91 +16,119 @@ export interface Prefixes<I = RDF.NamedNode> {
+    function literal(value: string | number, languageOrDatatype?: string | RDF.NamedNode): Literal;
+    function variable(value: string): Variable;
+    function defaultGraph(): DefaultGraph;
+    function quad(subject: RDF.Quad_Subject, predicate: RDF.Quad_Predicate, object: RDF.Quad_Object, graph?: Quad_Graph): Quad;
```
The `graph` param should be `RDF. Quad_Graph`.

LaurensRietveld replied:
Resolved :+1:

rubensworks commented:
@LaurensRietveld I also suggest bumping the RDFJS typings version to 2.0.

LaurensRietveld commented:
@rubensworks I've upped the rdfjs typings version to 2.0, and resolved some conversations about the statement discussions considering we've settled on the defaults for that.

rubensworks commented:
@LaurensRietveld I'm in the process of updating my libraries to these new typings, and I'm wondering if the I/O types of `triple` and `quad` in the DataFactory are necessary.

More concretely, are there cases where the input and output type will/can/may be different?
As far as I can see, we may be able to use only a single type parameter for this, which would simplify the usage a bit.

LaurensRietveld commented:
@rubensworks I'm assuming you're talking about the rdf-js typings here? In that case you're right, there aren't any uses there for having a different I/O typing. I'll draft a new PR for that one

rubensworks commented:
> I'm assuming you're talking about the rdf-js typings here?

Yes.

> I'll draft a new PR for that one

👍

LaurensRietveld commented:
Great, see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30164
